CDAP-7536 Program Lifecycle Documentation

awholegunch opened the pull request:
Running as a Quick Build: http://builds.cask.co/browse/CDAP-DQB195-11
Pages of interest: 
- http://builds.cask.co/artifact/CDAP-DQB195/shared/build-11/Docs-HTML/4.0.0-SNAPSHOT/en/developers-manual/building-blocks/program-lifecycle.html
- http://builds.cask.co/artifact/CDAP-DQB195/shared/build-11/Docs-HTML/4.0.0-SNAPSHOT/en/developers-manual/building-blocks/transaction-system.html
- http://builds.cask.co/artifact/CDAP-DQB195/shared/build-11/Docs-HTML/4.0.0-SNAPSHOT/en/examples-manual/examples/fileset-example.html#filesetservice
- http://builds.cask.co/artifact/CDAP-DQB195/shared/build-11/Docs-HTML/4.0.0-SNAPSHOT/en/examples-manual/examples/sport-results.html#uploadservice
- http://builds.cask.co/artifact/CDAP-DQB195/shared/build-11/Docs-HTML/4.0.0-SNAPSHOT/en/developers-manual/building-blocks/mapreduce-programs.html

Fix for:
- https://issues.cask.co/browse/CDAP-7531
- https://issues.cask.co/browse/CDAP-7536

Commit:
draft

Commit:
Documentation on lifecycle of programs.

anew commented on cdap-docs/developers-manual/source/building-blocks/program-lifecycle.rst:
```diff
@@ -0,0 +1,269 @@
+the same lifecycle. The exceptions are flows and services, as they have sub-programs
+(flowlets and service handlers) that handle part of the lifecycle.
+
+In order to be a CDAP program, a program must implement the ``ProgramLifecycle``
```
Not true. Unfortunately, we are no consistent about that. For example, MapReduce can optionally implement ProgramLifecycle<MapReduceContext>.

anew commented on cdap-docs/developers-manual/source/building-blocks/program-lifecycle.rst:
```diff
@@ -0,0 +1,269 @@
+(flowlets and service handlers) that handle part of the lifecycle.
+
+In order to be a CDAP program, a program must implement the ``ProgramLifecycle``
+interface. It requires the implementation of these methods:
```
ProgramLIfecycle only requires initialize() and destroy().

anew commented on cdap-docs/developers-manual/source/building-blocks/program-lifecycle.rst:
```diff
@@ -0,0 +1,269 @@
+- At runtime:
+
+  - ``initialize()``
+  - ``getConfig()``
```
Here, there is a distinction between the program and the abstract program class. The program is required to implement initialize(ProgramContext). The AbstractProgram implements that in such a way that it stores the context in an instance variable and then calls initialize() with no arguments. So, if a program extends from the abstract program class, it implements initialize() without arguments, and it can call getContext() to access the context saved by the abstract class. Similar for the config.

anew commented on cdap-docs/developers-manual/source/building-blocks/program-lifecycle.rst:
```diff
@@ -0,0 +1,269 @@
+Deployment
+==========
+At time of deployment, a program's ``configure()`` method is called. 
+It is called once, and produces an immutable program specification::
```
what you are describing now are the abstract program classes, which are convenience classes wrapping the actual program classes.

anew commented on cdap-docs/developers-manual/source/building-blocks/program-lifecycle.rst:
```diff
@@ -0,0 +1,269 @@
+
+Transactions
+============
+The relationship between transactions and lifecycle depends on which method is involved:
```
This needs rewording. Let's talk tomorrow.

anew commented on cdap-docs/developers-manual/source/building-blocks/program-lifecycle.rst:
```diff
@@ -0,0 +1,269 @@
+
+Dataset Access in MapReduce
+---------------------------
+Due to a limitation in the CDAP MapReduce implementation (see :cask-issue:`CDAP-6099`),
```
I think this should be mentioned in the known issues of the release notes, not the actual documentation.

anew reviewed (commented):
hey JJ, please see my comments. Perhaps we can chat tomorrow.

Commit:
Yarn -> YARN as appropriate.

Commit:
Correct usage and links to Apache Twill.

awholegunch commented:
"On-hold" as we have changes to come

Commit:
Fix formatting.

Commit:
Revisions; added link.

awholegunch commented:
Waiting for copy from Andreas, which this material will link to.

Commit:
Put table at bottom.

Commit:
Adding new section on transactions, and links for other sections to use.

Commit:
Fix formatting.

anew commented on cdap-docs/developers-manual/source/building-blocks/program-lifecycle.rst:
```diff
@@ -0,0 +1,188 @@
+transactions.
+
+**When an application is run,** each program of an application (and any sub-programs) is
+executed by calling first ``initilaize()`` and (eventually) ``destroy()`` for each program
```
typo: initialize

awholegunch replied:
fixed.

anew commented on cdap-docs/developers-manual/source/building-blocks/program-lifecycle.rst:
```diff
@@ -0,0 +1,188 @@
+executed by calling first ``initilaize()`` and (eventually) ``destroy()`` for each program
+and sub-program. In these methods, dataset operations and transactions are allowed.
+
+The following of the program lifecycle is a combination of CDAP's conventions and the implmentation
```
typo: implementation

awholegunch replied:
Fixed.

anew commented on cdap-docs/developers-manual/source/building-blocks/program-lifecycle.rst:
```diff
@@ -0,0 +1,188 @@
+and sub-program. In these methods, dataset operations and transactions are allowed.
+
+The following of the program lifecycle is a combination of CDAP's conventions and the implmentation
+of specific interfaces. They can be summarised as:
```
do we use british of american spelling? Since we call it "initialize", should it also be summarized?

awholegunch replied:
Fixed.

anew commented on cdap-docs/developers-manual/source/building-blocks/program-lifecycle.rst:
```diff
@@ -0,0 +1,188 @@
+
+The ``initialize()`` method is called once, at the start of the program. The ``destroy()``
+method is called once, at the end of program before it is shutdown. If there is any
+cleanup required, it can be specified in this method.
```
specified/implemented?

awholegunch replied:
Fixed: implemented.

anew commented on cdap-docs/developers-manual/source/building-blocks/program-lifecycle.rst:
```diff
@@ -0,0 +1,188 @@
+the object called at runtime. Because the result of the deployment stage is an immutable
+program specification, any local member variables set during deployment will not be
+available during runtime. This behavior can cause unexpected null-pointer exceptions. The
+solution is instead to set these as properties in a configuration object, which is
```
the properties are in the specification. For example, in a flowlet:     getContext().getSpecification().getProperties()

awholegunch replied:
Fixed.

anew commented on cdap-docs/developers-manual/source/building-blocks/program-lifecycle.rst:
```diff
@@ -0,0 +1,188 @@
+
+- ``destroy()`` Inside a transaction
+
+The exception to this are :ref:`Workers <workers-datasets>`, which are run inside their own transaction.
```
which have to execute their own, explicit transactions.

awholegunch replied:
Fixed.

anew commented on cdap-docs/developers-manual/source/building-blocks/program-lifecycle.rst:
```diff
@@ -0,0 +1,188 @@
+The exception to this are :ref:`Workers <workers-datasets>`, which are run inside their own transaction.
+See :ref:`workers and datasets <workers-datasets>` for details.
+
+Details on transactions in these methods is covered in the section on :ref:`using the
```
is/are

awholegunch replied:
Fixed.

anew commented on cdap-docs/developers-manual/source/building-blocks/program-lifecycle.rst:
```diff
@@ -0,0 +1,188 @@
+Details on transactions in these methods is covered in the section on :ref:`using the
+transaction system in programs <transaction-system-using-in-programs>`.
+
+Program Types
```
Because you have not mentioned any of the abstract program classes yet, they seem to come out of nowhere in the table that follows. I would add a small subsection on abstract programs: 

For convenience, most program types have a corresponding abstract program class. It is recommended to always extend abstract class instead of implementing the program interface. The abstract classes mainly provide:
- proxy methods for the program configurer's methods. 
- an initialize method that stores the program context in a class member and makes it available via getContext()
- a destroy method that does nothing.

awholegunch replied:
Added. (And adjusted.)

anew commented on cdap-docs/developers-manual/source/building-blocks/program-lifecycle.rst:
```diff
@@ -0,0 +1,188 @@
+
+   * - Program Type
+     - Sub-program Type
+     - Extends
```
Better to say "Abstract Class"

awholegunch replied:
Fixed.

anew commented on cdap-docs/developers-manual/source/building-blocks/transaction-system.rst:
```diff
@@ -158,3 +160,189 @@ It's important to note that the MapReduce framework will reattempt a task (Mappe
+tasks (mappers and reducers) and MapReduce helpers (such as partitioners and comparators)
+are always run inside a transaction: the long-running transaction that encapsulates an
+entire MapReduce job (see :ref:`above <transaction-system-transactions-mapreduce>`). 
+
```
Is this a good place to mention: Please see [link to spark section] for how to use transactions in Spark?

awholegunch replied:
Fixed.

anew commented on cdap-docs/developers-manual/source/building-blocks/transaction-system.rst:
```diff
@@ -158,3 +160,189 @@ It's important to note that the MapReduce framework will reattempt a task (Mappe
+
+For example, in the ``FileSetService`` of the :ref:`FileSetExample <examples-fileset>`::
+
+  @GET
```
after I merge my PR, you can replace this with a literal include...

awholegunch replied:
Will fix.

anew commented on cdap-docs/developers-manual/source/building-blocks/transaction-system.rst:
```diff
@@ -158,3 +160,189 @@ It's important to note that the MapReduce framework will reattempt a task (Mappe
+access the partition metadata, whereas the streaming of the file contents to the client is
+performed outside the transaction::
+
+  @GET
```
same here, can be converted into a literal include later.

awholegunch replied:
Will fix.

anew commented on cdap-docs/developers-manual/source/building-blocks/transaction-system.rst:
```diff
@@ -158,3 +160,189 @@ It's important to note that the MapReduce framework will reattempt a task (Mappe
+The name of the preference is ``system.data.tx.timeout``. 
+
+To configure the timeout for a sub-program (a flowlet or a custom workflow action), prefix
+the property name with ``flowlet.<name>`` or ``action.<name>``. For example, setting
```
preference name

awholegunch replied:
Fixed.

anew reviewed (commented):
Overall this looks very good now. Few typos and I have some minor suggestions.

Commit:
Fixes. To be reconciled later.

awholegunch commented:
"On-Hold" until reconciled with https://github.com/caskdata/cdap/pull/7399

anew commented:
LGTM now. Only the literal includes will change once https://github.com/caskdata/cdap/pull/7399 is merged.

anew commented:
7399 is now merged. If you rebase on release/4.0 and fix the literal includes, it is good to merge.

Commit:
Merge branch 'release/4.0' into docs/r4.0_7536_lifecycle

Commit:
Fix literal includes, MD5 hashes, and add links to examples of transactions.

Commit:
Adjust hash for BUILD.rst.

Commit:
Added a section on implicit transactions.

anew commented on cdap-docs/developers-manual/source/building-blocks/transaction-system.rst:
```diff
@@ -224,13 +224,26 @@ By default, CDAP will start an implicit transaction for these methods:
+.. literalinclude:: /../../../cdap-examples/HelloWorld/src/main/java/co/cask/cdap/examples/helloworld/HelloWorld.java
+   :language: java
+   :lines: 118-135
+   :dedent: 4
```
looks like this has to be 2. Otherwise you turn the "public" into "blic"

anew replied:
this still needs fixing?

anew commented on cdap-docs/developers-manual/source/building-blocks/transaction-system.rst:
```diff
@@ -224,13 +224,26 @@ By default, CDAP will start an implicit transaction for these methods:
+   :lines: 118-135
+   :dedent: 4
+
+For flowlet process methods, this starting of implicit transactions cannot be disabled,
```
This transition is a little abrupt. Perhaps you can add something like this:

In the following we will explain how to disable implicit transaction. First all of all, not that for flowlet process methods, ...

anew commented on cdap-docs/developers-manual/source/building-blocks/transaction-system.rst:
```diff
@@ -224,13 +224,26 @@ By default, CDAP will start an implicit transaction for these methods:
+For MapReduce programs, the lifecycle methods of MapReduce tasks (mappers and reducers)
+and MapReduce helpers (such as partitioners and comparators) are always run inside a
+transaction: the long-running transaction that encapsulates an entire MapReduce job (see
+:ref:`above <transaction-system-transactions-mapreduce>`).
```
I think that should be the next section.

Commit:
Fileset example: changes context discussing reclaiming datasets.
Datasets overview: Added comments to "discarding datasets" section
Tx System: added an implicit tx example (HelloWorld), moved the entire
"Using Transactions in Programs" to precede "Levels of Conflict Detection"
with a minor change (a link label for MapReduce, removing usage of "above")

anew commented on cdap-docs/developers-manual/source/building-blocks/transaction-system.rst:
```diff
@@ -335,9 +254,94 @@ the timeout in seconds to the ``execute()`` method::
+
+.. _transaction-system-transactions-mapreduce:
+
+Transactions and MapReduce
```
Can we put this after "Using Transactions in Programs"? I almost feel like this should even be moved into the MapReduce chapter. Between "MapReduce and Datasets" and "MapReduce and Resources". And add a link to it here, same as we did for Spark.

awholegunch replied:
Fixed. `dedent`. Moved section to datasets.

Commit:
Moved sections, fixed links, fixed dedent.

anew commented:
LGTM
